libc/include/sys/queue.h

```diff
@@ -96,27 +148,61 @@
     struct type *next;                                                         \
   }
 
+#define STAILQ_CLASS_ENTRY(type)                                               \
+  struct {                                                                     \
+    class type *next;                                                          \
+  }
+
 // Singly-linked tail queue access methods.
 
 #define STAILQ_EMPTY(head) ((head)->first == NULL)
 #define STAILQ_FIRST(head) ((head)->first)
+#define STAILQ_LAST(head, type, field)                                         \
+  STAILQ_EMPTY(head)                                                           \
+  ? NULL                                                                       \
+  : (QUEUE_TYPEOF(type) *)((char *)(head)->last -                              \
+                           offsetof(QUEUE_TYPEOF(type), field))
 #define STAILQ_NEXT(elem, field) ((elem)->field.next)
 
 #define STAILQ_FOREACH(var, head, field)                                       \
   for ((var) = STAILQ_FIRST(head); (var); (var) = STAILQ_NEXT(var, field))
 
+#define STAILQ_FOREACH_FROM(var, head, field)                                  \
+  for ((var) = ((var) ? (var) : STAILQ_FIRST(head)); (var);                    \
+       (var) = STAILQ_NEXT(var, field))
+
 #define STAILQ_FOREACH_SAFE(var, head, field, tvar)                            \
   for ((var) = STAILQ_FIRST(head);                                             \
        (var) && ((tvar) = STAILQ_NEXT(var, field), 1); (var) = (tvar))
 
+#define STAILQ_FOREACH_SAFE(var, head, field, tvar)                            \
+  for ((var) = ((var) ? (var) : STAILQ_FIRST(head));                           \
+       (var) && ((tvar) = STAILQ_NEXT(var, field), 1); (var) = (tvar))
```

Comment:
accidental redefinition? Should this be `STAILQ_FOREACH_FROM_SAFE`? (missing `FROM_`)

Reply:
Done.